File: src/main/java/com/eightbitforest/termu/gui/term/MainWindow.java
package main.java.com.eightbitforest.termu.gui.term;

import main.java.com.eightbitforest.termu.gui.term.terminal.ITerminal;
import main.java.com.eightbitforest.termu.gui.term.terminal.LanternaTerminal;
import main.java.com.eightbitforest.termu.gui.term.terminal.pixelterminal.BraillePixelTerminal;
import main.java.com.eightbitforest.termu.gui.term.terminal.pixelterminal.IPixelTerminal;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;

class MainWindow {
    void start() {
        ITerminal terminal = new LanternaTerminal();
        IPixelTerminal pixelTerminal = new BraillePixelTerminal();

        terminal.init();
        pixelTerminal.init(256, 224);

        int[][][] allPixels = new int[1][][];
        try {
            allPixels[0] = getPixels(ImageIO.read(new File("test1.png")));
        } catch (IOException e) {
            e.printStackTrace();
        }

        int frames = 0;
        int i = 0;
        long time = System.nanoTime();
        while (true) {
            for (int x = 0; x < 256; x++) {
                for (int y = 0; y < 224; y++) {
                    pixelTerminal.setPixel(x, y, allPixels[i][x][y]);
                }
            }

            pixelTerminal.render(terminal);
            terminal.refresh();

            i++;
            if (i >= allPixels.length)
                i = 0;

            frames++;
            if (System.nanoTime() - time >= 1e+9){
                System.out.println("FPS: " + frames);
                frames = 0;
                time = System.nanoTime();
            }
        }
    }

    private static int[][] getPixels(BufferedImage image) {
        int w = image.getWidth();
        int h = image.getHeight();

        int[][] pixels = new int[w][h];

        for (int x = 0; x < w; x++) {
            for (int y = 0; y < h; y++) {
                pixels[x][y] = image.getRGB(x, y);
            }
        }

        return pixels;
    }
}

File: src/gui/term/windows/MainWindow.cpp
#include "MainWindow.h"
#include "../terminal/CursesTerminal.h"
#include "../terminal/pixelterminal/BraillePixelTerminal.h"
#include <opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui.hpp>

#include <iostream>

void MainWindow::Start() {
    Terminal* terminal = new CursesTerminal();
    PixelTerminal* pixelTerminal = new BraillePixelTerminal();

    terminal->Init();
    pixelTerminal->Init(256, 224);

    int numImages = 2;
    int ***allPixels = new int **[numImages];

    allPixels[0] = GetPixels("./test1.png");
    allPixels[1] = GetPixels("./test2.png");

    int frames = 0;
    int i = 0;

    auto time = std::chrono::high_resolution_clock::now();
    while (true) {
        for (int x = 0; x < 256; x++) {
            for (int y = 0; y < 224; y++) {
                pixelTerminal->SetPixel(x, y, allPixels[i][x][y]);
            }
        }

        pixelTerminal->Render(terminal);
        terminal->Refresh();

        i++;
        if (i >= numImages)
            i = 0;

        frames++;
        if (std::chrono::duration_cast<std::chrono::seconds>(
                std::chrono::high_resolution_clock::now() - time).count() >= 1) {
            std::cout << "FPS: " << frames << std::endl;
            frames = 0;
            time = std::chrono::high_resolution_clock::now();
        }
    }

    terminal->Close();
}

int **MainWindow::GetPixels(std::string file) {
    cv::Mat image;
    image = cv::imread(file);

    int w = image.cols;
    int h = image.rows;

    int **pixels = new int *[w];

    for (int x = 0; x < w; x++) {
        pixels[x] = new int[h];
        for (int y = 0; y < h; y++) {
            cv::Vec3b rgb = image.at<cv::Vec3b>(y, x);
            pixels[x][y] = (rgb[0] << 16) | (rgb[1] << 8) | (rgb[2]);
        }
    }

    return pixels;
}

